### DISCUSSION TITLE
UART communication to Raspberry Pi zero 2 w

### DISCUSSION THREAD

**Discussion by mfxjx000205** [bitcraze/discussions#1091] [Q&A]
Hi , bitcraze,

I had written earlier to ask about assembling the crazybolt and all went well. For my thesis I wanted an onboard computer on top of crazyflie/crazybolt, and the **Raspberry Pi zero 2w** was my choice, they need to fly together. Based on this, I have the following questions:

Q1. I've followed everything in the "UART communication" blog to configure the Raspberry Pi and crazyflie, and I've connected the crazyflie to the raspberry pi correctly using DuPont cables (I'm not using crazybolt yet). However, the UART connection does not work anyway. There are two main cases.

case1：Cannot connect at all, displaying the following. For most of the time I've tried, it's been like this.

`root@raspberrypi:~# python python-test/2testuart.py 
INFO:cflib.crazyflie:Callback->Connection initialized[serial://ttyAMA0]
INFO:cflib.crtp.serialdriver:Initialized serial driver.
CPX UART transport
Connecting to UART on /dev/ttyAMA0 @ 576000...`

case2：Can connect, but cannot communicate correctly. On several consecutive occasions during the experiment, it was possible to connect, but could not communicate correctly. The error is reported as in the picture, and after I restarted the Raspberry Pi once, it became case 1 again.

![20231215-133043](https://github.com/bitcraze/discussions/assets/32474853/1f8ea11a-1abd-461a-9a08-285b5983e368)

Q2：Can I ask you to analyze what are the advantages and disadvantages of using UART connection and using USB connection each. Or more to the point, why is it not better to connect directly to crazyflie on the fly via USB. I don't need to use PA controls and can connect via USB and have it work fine.

Best Regards,

Hanjie From Southeast University.

**Comment by knmcguire** [comment]
Hi! Which blogpost in particular did you look at? It might be quite old and no longer valid. Was it this one? https://www.bitcraze.io/2017/07/crazyflie-based-quadcopter-with-raspberrypi-camera/

Or did you use these instructions: https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/development/uart_communication/

The uart connection of the raspberry pi is not the most stable and there are some ifs and buts. I don't know if you looked at other discussions before like this one: https://github.com/orgs/bitcraze/discussions/473

If you have the possibility to a usb connection that probably has the preference for my part in terms of stability. However, uart 'might' be faster so if you are running a control loop through it then that might be better, but it does not seem like you'll need it for that right?
